File: December 2024 Leetcode Solution/December-14.py
from collections import deque
from typing import List

class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        start = 0
        total_subarrays = 0
        max_deque, min_deque = deque(), deque()
        for end, num in enumerate(nums):
            while max_deque and nums[max_deque[-1]] <= num:
                max_deque.pop()
            max_deque.append(end)
            while min_deque and nums[min_deque[-1]] >= num:
                min_deque.pop()
            min_deque.append(end)
            while nums[max_deque[0]] - nums[min_deque[0]] > 2:
                if max_deque[0] == start:
                    max_deque.popleft()
                if min_deque[0] == start:
                    min_deque.popleft()
                start += 1
            total_subarrays += end - start + 1

        return total_subarrays

File: December 2024 Leetcode Solution/December-14.rs
impl Solution {
    pub fn continuous_subarrays(nums: Vec<i32>) -> i64 {
        let mut start = 0;
        let mut total_subarrays = 0i64;
        let mut max_deque: std::collections::VecDeque<usize> = std::collections::VecDeque::new();
        let mut min_deque: std::collections::VecDeque<usize> = std::collections::VecDeque::new();
        for (end, &num) in nums.iter().enumerate() {
            while let Some(&last) = max_deque.back() {
                if nums[last] <= num {
                    max_deque.pop_back();
                } else {
                    break;
                }
            }
            max_deque.push_back(end);
            while let Some(&last) = min_deque.back() {
                if nums[last] >= num {
                    min_deque.pop_back();
                } else {
                    break;
                }
            }
            min_deque.push_back(end);
            while nums[*max_deque.front().unwrap()] - nums[*min_deque.front().unwrap()] > 2 {
                if *max_deque.front().unwrap() == start {
                    max_deque.pop_front();
                }
                if *min_deque.front().unwrap() == start {
                    min_deque.pop_front();
                }
                start += 1;
            }
            total_subarrays += (end - start + 1) as i64;
        }

        total_subarrays
    }
}
